onnxruntime/core/providers/xnnpack/nn/unpool.cc

```diff
@@ -0,0 +1,266 @@
+
+  // For feature dims calculate reversing the formula used for MaxPool
+  for (size_t dim = 0; dim < pool_attrs.kernel_shape.size(); ++dim) {
+    inferred_output_dims[dim + 1] =
+        (input_shape[dim + 1] - 1) * pool_attrs.strides[dim] -
+        (pool_attrs.pads[dim] + pool_attrs.pads[pool_attrs.kernel_shape.size() + dim]) +
+        pool_attrs.kernel_shape[dim];
+  }
+
+  return inferred_output_dims;
+}
+}  // namespace
+
+// MaxUnpool doesn't have any quantization params
+bool MaxUnpool::IsOnnxNodeSupported(const NodeUnit& node_unit,
+                                    const GraphViewer& graph_viewer) {
+  ORT_UNUSED_PARAMETER(graph_viewer);
+  bool supported = false;
+  const onnxruntime::Node& node = node_unit.GetNode();
+  // use do {} while(false) so it's easier to set a breakpoint on the return
+  do {
+    // MaxUnpool has 2-3 inputs.
+    auto input_defs = node.InputDefs();
+
+    if (input_defs.size() == 3) {
+      const auto* output_shape = graph_viewer.GetConstantInitializer(input_defs[2]->Name(), true);
+      if (output_shape == nullptr || output_shape->dims_size() != 1 || output_shape->dims(0) != 4) {
+        break;
+      }
+    }
+    const auto& x_arg = *input_defs[0];
+
+    const auto* x_type = x_arg.TypeAsProto();
+    if (x_type == nullptr ||
+        (x_type->tensor_type().elem_type() != ONNX_NAMESPACE::TensorProto_DataType_FLOAT)) {
+      break;
+    }
+
+    // we only support 2D (4 dims with batch and channel)
+    const auto* x_shape = x_arg.Shape();
+    if (!x_shape || x_shape->dim_size() != 4) {
+      break;
+    }
+
+    // require C, H, W to be known so we can construct the xnnpack kernel prior to Compute
+    if (!x_shape->dim(1).has_dim_value() ||
+        !x_shape->dim(2).has_dim_value() ||
+        !x_shape->dim(3).has_dim_value()) {
+      break;
+    }
+
+    ProtoHelperNodeContext nc(node);
+    OpNodeProtoHelper info(&nc);
+    PoolAttributes pool_attrs(info, "MaxUnpool", node.SinceVersion());
+
+    if (!IsPaddingTypeSupported(pool_attrs.auto_pad)) {
+      break;
+    }
+
+    supported = true;
+  } while (false);
+
+  return supported;
+}
+
+MaxUnpool::MaxUnpool(const OpKernelInfo& info)
+    : XnnpackKernel(info),
+      pool_attrs_{info, "MaxUnpool", info.node().SinceVersion()} {
+  num_inputs_ = OpKernel::Node().InputDefs().size();
+  info.GetAttrOrDefault<int64_t>("mode", &is_indice_produced_by_xnnpack_, 0);
+  // input is NHWC and we only support input with 4 dims. we checked C, H, W were all known in the op support checker
+  const auto& X_arg = *Node().InputDefs()[0];
+  auto X_shape = utils::GetTensorShapeFromTensorShapeProto(*X_arg.Shape());
+
+  int64_t H = X_shape[1];
+  int64_t W = X_shape[2];
+  int64_t C = X_shape[3];
+
+  // create NCHW shape to calculate most of the output shape. 'N' is set in Compute.
+  TensorShapeVector input_shape{1, C, H, W};
+  auto pads = pool_attrs_.pads;
+  output_dims_ = InferOutputSizeForUnPool(pool_attrs_, X_shape);
```

Comment:
Given InferOutputSizeForUnPool is a local function can we update it to work with NHWC input?

Reply:
Not quite understand. actually, This function only works with NHWC input

Reply:
MaxUnpool is layout sensitive